sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,425 @@
+
+        let (sui_amount, _, lsp_supply) = get_amounts(pool);
+
+        let sui_added = Balance::value(&sui_balance);
+        let share_minted = (sui_added * lsp_supply) / sui_amount;
+
+        Balance::join(&mut pool.sui, sui_balance);
+        Balance::join(&mut pool.token, token_balance);
+
+        Coin::mint(share_minted, &mut pool.lsp_treasury, ctx)
+    }
+
+    /// Entrypoint for the `remove_liquidity` method. Transfers
+    /// withdrawn assets to the sender.
+    public(script) fun remove_liquidity_<T>(
+        pool: &mut Pool<T>,
+        lsp: Coin<LSP<T>>,
+        ctx: &mut TxContext
+    ) {
+        let (sui, token) = remove_liquidity(pool, lsp, ctx);
+        let sender = TxContext::sender(ctx);
+
+        Transfer::transfer(sui, sender);
+        Transfer::transfer(token, sender);
+    }
+
+    /// Remove liquidity from the `Pool` by burning `Coin<LSP>`.
+    /// Returns `Coin<T>` and `Coin<SUI>`.
+    public fun remove_liquidity<T>(
+        pool: &mut Pool<T>,
+        lsp: Coin<LSP<T>>,
+        ctx: &mut TxContext
+    ): (Coin<SUI>, Coin<T>) {
+        let lsp_amount = Coin::value(&lsp);
+
+        assert!(lsp_amount > 0, EZeroAmount);
+
+        let (sui_amt, tok_amt, lsp_supply) = get_amounts(pool);
+        let sui_removed = (sui_amt * lsp_amount) / lsp_supply;
+        let tok_removed = (tok_amt * lsp_amount) / lsp_supply;
+
+        Coin::burn(lsp, &mut pool.lsp_treasury);
+
+        (
+            Coin::withdraw(&mut pool.sui, sui_removed, ctx),
+            Coin::withdraw(&mut pool.token, tok_removed, ctx)
+        )
+    }
+
+    /// Public getter for the price of SUI in token T.
+    /// - How much SUI one will get if they send `to_sell` amount of T;
+    public fun sui_price<T>(pool: &Pool<T>, to_sell: u64): u64 {
+        let (sui_amt, tok_amt, _) = get_amounts(pool);
+        get_input_price(to_sell, tok_amt, sui_amt, pool.fee_percent)
+    }
+
+    /// Public getter for the price of token T in SUI.
+    /// - How much T one will get if they send `to_sell` amount of SUI;
+    public fun token_price<T>(pool: &Pool<T>, to_sell: u64): u64 {
+        let (sui_amt, tok_amt, _) = get_amounts(pool);
+        get_input_price(to_sell, sui_amt, tok_amt, pool.fee_percent)
+    }
+
+
+    /// Get most used values in a handy way:
+    /// - amount of SUI
+    /// - amount of token
+    /// - total supply of LSP
+    public fun get_amounts<T>(pool: &Pool<T>): (u64, u64, u64) {
+        (
+            Balance::value(&pool.sui),
+            Balance::value(&pool.token),
+            Coin::total_supply(&pool.lsp_treasury)
+        )
+    }
+
+    /// Calculate the output amount minus the fee - 0.3%
+    fun get_input_price(
+        input_amount: u64, input_reserve: u64, output_reserve: u64, fee_percent: u64
+    ): u64 {
+        let input_amount_with_fee = input_amount * (1000 - fee_percent); // 0.3% fee
+        let numerator = input_amount_with_fee * output_reserve;
```

Comment:
This might easily overflow if both values are > u32Max (possible even if one of them is > u32Max and their product happens to be > u64Max). This practically makes our implementation supporting values up to the u32 limit. If you ask me, I believe we should soon support u128 or better u256 for defi products, u32/u64 will be proven in practice too small for many asset types.

In any case, please add a unit test transferring large values (+ reserve amount to be > u32Max). Also tests that we transfer zero.

Reply:
For the math operations I've added upcasts to u128; in the end result is casted back to u64.